_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+
+``` jsx
+handleTitleClick() {
+  this.setState(prevState => ({ active: !prevState.active }));
+};
+```
+
+Now our onClick handler properly deals with existing state.
+
+The onClick handler can only update state by accessing the `this` context. This means we'll also need to bind it to `this` properly within the constructor.
+
+``` jsx
+constructor(props) {
+  super(props);
+
+  this.state = {
+    active: false,
+  };
+
+  this.handleTitleClick = this.handleTitleClick.bind(this);
+}
+```
+<blockquote class="Info Info-right">
+  "If you are using the experimental public class fields syntax, you can use class fields to correctly bind callbacks"
+  <cite>- <a href="https://reactjs.org/docs/handling-events.html">reactjs.org</a></cite>
+</blockquote>
+
+While this currently works, we can clean this up by taking it one step further using the **public class fields syntax**.
+
+I really love how easy this syntax reads and how it cleans up your constructor function from all the binding cruft.
+
+``` jsx
+constructor(props) {
+  super(props);
+
+  this.state = {
+    active: false,
+  };
+
+  // No binding necessary
+}
+
+// Instead of () {} is now = () => {}
+handleTitleClick = () => {
+  this.setState({ active: !this.state.active });
+}
+```
+
+So with all that out of the way, we now have our first iteration of our component below.
+
+``` jsx
+import React from 'react';
+
+class Drawer extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      active: false,
+    };
+  }
+
+  handleTitleClick = () => {
+    this.setState(prevState => ({ active: !prevState.active }));
+  };
+
+  render() {
+    const { title, children } = this.props;
+    const drawerStyles = this.state.active ? 'is-expanded' : '';
+
+    return (
+      <div className={`Drawer ${drawerStyles}`}>
+        <div
+          className="Drawer-title"
+          onClick={this.handleTitleClick}
+        >
+          {title}
+        </div>
+        <div className="Drawer-content">
+          {children}
+        </div>
+      </div>
+    );
+  }
+}
+
+export default Drawer;
+
+/////////////////////////////////////
+// And here's how you would use it //
+/////////////////////////////////////
+import React from 'react';
+import Drawer from '~/components/Drawer'; // The location of our new component
+
+class AnotherComponent extends React.Component {
+  // code omitted
+  render() {
+    return (
+      <div className="AnotherComponent">
+        <h1>Welcome to my component</h1>
+        <Drawer title="My Drawer Title">
+          <ul>
+            <li>Item 1</li>
+            <li>Item 2</li>
+            <li>Item 3</li>
+          </ul>
+        </Drawer>
+      </div>
+    );
+  }
+}
+```
+
+Here's what it looks like so far. Really the only thing it does is provide
+structure and toggle a CSS class `is-expanded` based on state.
+
```

Comment:
and toggle a CSS class `is-expanded` based on state. <<<this seems like it might have a grammar error in it but i can't tell for sure.